test/test_utils.hh

```diff
@@ -29,19 +80,25 @@ void matrix_iterator(
 
     int64_t B_mt = B.mt();
     int64_t B_nt = B.nt();
-    if constexpr (std::is_same<matrix_type, slate::Matrix<scalar_t>>::value) {
 
-        int64_t A_j = 0, A_jj = 0;
-        for (int64_t B_j = 0; B_j < B_nt; ++B_j) {
+    constexpr bool is_general = std::is_same_v<matrix_type, slate::Matrix<scalar_t>>;
+    assert( is_general == (A.uplo() == slate::Uplo::General) );
+    bool is_upper = (A.uplo() == slate::Uplo::Upper);
+
+    int64_t A_j = 0, A_jj = 0;
+    for (int64_t B_j = 0; B_j < B_nt; ++B_j) {
 
-            int64_t A_i = 0, A_ii = 0;
-            for (int64_t B_i = 0; B_i < B_mt; ++B_i) {
+        int64_t A_i = 0, A_ii = 0;
+        for (int64_t B_i = 0; B_i < B_mt; ++B_i) {
+
+            if (is_general || (is_upper ? (B_i <= B_j) : (B_i >= B_j))) {
                 #pragma omp task shared(A, B) \
                                  firstprivate( B_i, B_j, A_i, A_j, A_ii, A_jj )
                 {
+                    int tag = A_i + A_j * A.mt();
```

Comment:
MPI tags are only guaranteed up to 32768. [Intel supports up to 2^20 = 1 Mi](https://github.com/Nek5000/Nek5000/issues/65#issuecomment-240175658). For a square matrix, that allows (2^20)^(0.5) = 2^10 = 1024 block-rows and block cols. [Cray supposedly supports 2^21](https://github.com/Nek5000/Nek5000/issues/65#issuecomment-240175658).

It would be good to query MPI_TAG_UB and compare to mt*nt, at least throwing an error if it is exceeded.

Reply:
Added to TODO items #156.